java/client/src/main/java/glide/connectors/handlers/CallbackDispatcher.java

```diff
@@ -58,21 +64,56 @@ public CompletableFuture<Response> registerConnection() {
     }
 
     /**
-     * Complete the corresponding client promise and free resources.
+     * Complete the corresponding client promise, handle error and free resources.
      *
      * @param response A response received
      */
     public void completeRequest(Response response) {
+        if (response.hasClosingError()) {
+            // According to https://github.com/aws/glide-for-redis/issues/851
+            // a response with a closing error may arrive with any/random callback ID (usually -1)
+            // CommandManager and ConnectionManager would close the UDS channel on ClosingException
+            responses
+                    .values()
+                    .forEach(f -> f.completeExceptionally(new ClosingException(response.getClosingError())));
+            responses.clear();
+            return;
+        }
         // Complete and return the response at callbackId
         // free up the callback ID in the freeRequestIds list
         int callbackId = response.getCallbackIdx();
         CompletableFuture<Response> future = responses.remove(callbackId);
-        freeRequestIds.add(callbackId);
         if (future != null) {
+            freeRequestIds.add(callbackId);
+            if (response.hasRequestError()) {
+                RequestError error = response.getRequestError();
+                String msg = error.getMessage();
+                switch (error.getType()) {
+                    case Unspecified:
+                        // Unspecified error on Redis service-side
+                        future.completeExceptionally(new RequestException(msg));
+                    case ExecAbort:
+                        // Transactional error on Redis service-side
+                        future.completeExceptionally(new ExecAbortException(msg));
+                    case Timeout:
+                        // Timeout from Glide to Redis service
+                        future.completeExceptionally(new TimeoutException(msg));
+                    case Disconnect:
+                        // Connection problem between Glide and Redis
+                        future.completeExceptionally(new ConnectionException(msg));
+                    default:
+                        // Request or command error from Redis
+                        future.completeExceptionally(new RequestException(msg));
+                }
+            }
             future.completeAsync(() -> response);
         } else {
             // TODO: log an error.
             // probably a response was received after shutdown or `registerRequest` call was missing
+
+            System.err.printf(
+                    "Received a response for not registered callback id %d, request error = %s%n",
```

Comment:
Is this supposed to be \n instead of %n?
```suggestion
                    "Received a response for not registered callback id %d, request error = %s\n",
```

Reply:
Not a mistake - `%n` is rendered as `\n`. Thank you for proofreading though.